src/main/java/me/lokka30/treasury/plugin/command/treasury/subcommand/migrate/AccountMigrator.java

```diff
@@ -0,0 +1,83 @@
+package me.lokka30.treasury.plugin.command.treasury.subcommand.migrate;
+
+import me.lokka30.treasury.api.economy.EconomyProvider;
+import me.lokka30.treasury.api.economy.account.Account;
+import me.lokka30.treasury.api.economy.currency.Currency;
+import me.lokka30.treasury.api.economy.response.EconomyException;
+import me.lokka30.treasury.api.economy.response.EconomySubscriber;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+
+interface AccountMigrator<T extends Account> {
+
+    @NotNull String getBulkFailLog(@NotNull Throwable throwable);
+
+    @NotNull String getInitLog(@NotNull UUID uuid);
+
+    @NotNull String getErrorLog(@NotNull UUID uuid, @NotNull Throwable throwable);
+
+    @NotNull BiConsumer<@NotNull EconomyProvider, @NotNull EconomySubscriber<Collection<UUID>>> requestAccountIds();
+
+    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull UUID, @NotNull EconomySubscriber<T>> requestAccount();
+
+    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull UUID, @NotNull EconomySubscriber<Boolean>> checkAccountExistence();
+
+    @NotNull TriConsumer<@NotNull EconomyProvider, @NotNull UUID, @NotNull EconomySubscriber<T>> createAccount();
+
+    default void migrate(
+            @NotNull Phaser phaser,
+            @NotNull T fromAccount,
+            @NotNull T toAccount,
+            @NotNull MigrationData migration) {
+        for (Map.Entry<Currency, Currency> fromToCurrency : migration.migratedCurrencies().entrySet()) {
+            Currency fromCurrency = fromToCurrency.getKey();
+            Currency toCurrency = fromToCurrency.getValue();
+            CompletableFuture<Double> balanceFuture = new CompletableFuture<>();
```

Comment:
think this logic can be migrated to EconomySubscriber.asFuture

Reply:
It could be, though not directly. I'm reluctant to do that without doing a full pass over the entirety of migration, which I don't really have time to do in the coming week - life got busy, unfortunately.

Reply:
gotcha.

Reply:
I don't see why need this. Since what we have works now and Jikoo is busy we may as well do it in another pr so we can close this one and continue working on other things.